src/avm2/native.ts

```diff
@@ -1063,14 +1063,17 @@ module Shumway.AVM2.AS {
       if (re === (void 0) || re === null) {
         return null;
       } else {
-        if (re instanceof XRegExp && re.global) {
-          var matches = [], m;
-          while ((m = re.exec(this))) {
-            matches.push(m[0]);
+        if (re instanceof ASRegExp) {
+          if (re.global) {
+            var matches = [], m;
+            while ((m = re.exec(<string><any>this))) {
+                matches.push(m[0]);
+            }
+            return matches;
           }
-          return matches;
+          return re.match(<string><any>this);
         }
-        if (!(re instanceof XRegExp) && !(typeof re === 'string')) {
+        if (typeof re !== 'string') {
           re = String(re);
```

Comment:
Pre-existing, but this should be `asCoerceString`. And I would remove the `if` and just do the coercion.